pkg/services/baremetal/baremetal/baremetal.go

```diff
@@ -363,71 +365,126 @@ func (s *Service) getAssociatedHost(ctx context.Context) (*infrav1.HetznerBareMe
 	return &host, helper, nil
 }
 
-func (s *Service) chooseHost(ctx context.Context) (*infrav1.HetznerBareMetalHost, *patch.Helper, error) {
+// chooseHost tries to find a free hbmh.
+// If no hbmh was found, then hbmh and err are nil, and the string
+// "reason" contains human readable details.
+func (s *Service) chooseHost(ctx context.Context) (
+	hbmh *infrav1.HetznerBareMetalHost, patchHelper *patch.Helper, reason string, err error,
+) {
 	// get list of hosts scoped to namespace of machine
 	hosts := infrav1.HetznerBareMetalHostList{}
 	opts := &client.ListOptions{
 		Namespace: s.scope.BareMetalMachine.Namespace,
 	}
 
 	if err := s.scope.Client.List(ctx, &hosts, opts); err != nil {
-		return nil, nil, fmt.Errorf("failed to list hosts: %w", err)
+		return nil, nil, "", fmt.Errorf("failed to list hosts: %w", err)
 	}
 
 	labelSelector := s.getLabelSelector()
 
-	availableHosts := make([]*infrav1.HetznerBareMetalHost, 0, len(hosts.Items))
+	usableHosts := make([]*infrav1.HetznerBareMetalHost, 0, len(hosts.Items))
 
+	mapOfSkipReasons := make(map[string]int)
+
+	availableDevices := 0
```

Comment:
hosts not devices, and don't mix available and usable.. You renamed something from "available" to "usable" but here you name your variable "available" again. We should stick to one!

Reply:
We have three sets:

set 1: the big list which contains all devices.

set 2: Subset of set1. we removed the devices which are in use. This is called "available"

set 3: Subset of set2. We removed the one that don't match (like raid is wrong, labels don't match). This is called "usable".

I changed the name because I think it makes sense that there are more available devices than usable devices.

Does this make sense to you? Should I add a comment to the code?

Reply:
Ah I actually did not understand it. Maybe you can define "available" then before "usable" and quickly comment on it. 
However, it is still not "devices" but "hosts" ;)